app/src/main/java/org/schabi/newpipe/player/Player.java

```diff
@@ -1187,6 +1193,54 @@ private void initThumbnail(final String url) {
                 .loadImage(url, ImageDisplayConstants.DISPLAY_THUMBNAIL_OPTIONS, this);
     }
 
+    public void updateEndScreenThumbnail() {
+        if (currentThumbnail == null) {
+            return;
+        }
+
+        final float endScreenHeight;
+
+        updateScreenSize(); // ensure that screenHeight is initialized and thus not 0
+
+        // Show at least stream title and content creator on TVs and tablets (using the land layout)
+        // 85dp for height of R.id.detail_root and R.id.detail_title_root_layout
+        // 15/16sp for text size of video title
+        if (DeviceUtils.isTv(context) && !isFullscreen) {
+            final int videoInfoHeight =
+                    DeviceUtils.dpToPx(85, context) + DeviceUtils.spToPx(16, context);
+            endScreenHeight = Math.min(currentThumbnail.getHeight(),
+                    screenHeight - videoInfoHeight);
+        } else if (DeviceUtils.isTablet(context) && !isFullscreen) {
+            final int videoInfoHeight =
+                    DeviceUtils.dpToPx(85, context) + DeviceUtils.spToPx(15, context);
+            endScreenHeight = Math.min(currentThumbnail.getHeight(),
+                    screenHeight - videoInfoHeight);
+        } else { // fullscreen player: max height is the device height
+            endScreenHeight = Math.min(currentThumbnail.getHeight(), screenHeight);
+        }
+
+        // scale the player audio / end screen thumbnail
+        // This is necessary when the thumbnail's height is larger than the device's height
+        // and thus is enlarging the player's height
+        // causing the bottom playback controls to be out of the visible screen.
+        final Bitmap endScreenBitmap = Bitmap.createScaledBitmap(
+                currentThumbnail,
+                (int) (currentThumbnail.getWidth()
+                        / (currentThumbnail.getHeight() / endScreenHeight)),
+                (int) endScreenHeight,
+                true);
```

Comment:
conversion to int (rounding down) won't cause problems?

Reply:
no. The size is in pixels. The device cannot display an image with e.g. with a width of `800.8px`.  We want to scale the thumbnail to not make the player larger than the screen, so to be sure we do not make the player grow out of the display, we need to scale down.